### PR TITLE
8236736: Change notproduct JVM flags to develop flags

### PR DESCRIPTION
**Pull request by coleenp**
Remove the notproduct distinction for command line options, rather than trying to wrestle the macros to fix the bug that they've been treated as develop options for some time now.  This simplifies the command line option macros.

Tested with tier1-4, tier1 on Oracle platforms.  Also built shenandoah.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8236736](https://bugs.openjdk.org/browse/JDK-8236736): Change notproduct JVM flags to develop flags (**Enhancement** - P4)

### Reviewers
 * [Ioi Lam](https://openjdk.org/census#iklam) (@iklam - **Reviewer**) ⚠️ Review applies to [19b8f6b6](https://git.openjdk.org/jdk/pull/18541/files/19b8f6b673ac1483f4d73f61c72f66492ce372ca)
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**) ⚠️ Review applies to [00a241d3](https://git.openjdk.org/jdk/pull/18541/files/00a241d314f11a83cd5b58c83ab76727852a2056)
 * [Kim Barrett](https://openjdk.org/census#kbarrett) (@kimbarrett - **Reviewer**) ⚠️ Review applies to [00a241d3](https://git.openjdk.org/jdk/pull/18541/files/00a241d314f11a83cd5b58c83ab76727852a2056)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/18541/head:pull/18541` \
`$ git checkout pull/18541`

Update a local copy of the PR: \
`$ git checkout pull/18541` \
`$ git pull https://git.openjdk.org/jdk.git pull/18541/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 18541`

View PR using the GUI difftool: \
`$ git pr show -t 18541`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/18541.diff">https://git.openjdk.org/jdk/pull/18541.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/18541#issuecomment-2032032893)

### CODE REVIEW

**Review by iklam — CHANGES_REQUESTED** [review]
LGTM.

For the past 15 years, "notproduct" flags haven't been working as they claim to be in globals.hpp. That doesn't seem to have bothered anyone. This definitely looks like a design that no one needs and should be removed for simplcity.

There are some references to "notproduct" in test/hotspot/jtreg/runtime/CommandLine that need to be removed.

**Comment by coleenp** [comment]
Thank you for pointing out that I missed cleaning up the tests.  One failed but I didn't see it.

**Review by iklam — APPROVED** [review]
LGTM

File: src/hotspot/share/runtime/arguments.cpp [hunk 1]
```diff
@@ -3417,19 +3416,15 @@ bool Arguments::handle_deprecated_print_gc_flags() {
 
 static void apply_debugger_ergo() {
 #ifndef PRODUCT
-  // UseDebuggerErgo is notproduct
```
Comment (stefank) [inline src/hotspot/share/runtime/arguments.cpp:3420]: Now that the flag has been changed to a develop flag, it seems wrong that these are guarded by "#ifndef PRODUCT". Shouldn't this be changed to check for ASSERT instead?

File: src/hotspot/share/runtime/flags/jvmFlag.hpp [hunk 2]
```diff
@@ -115,7 +114,6 @@ class JVMFlag {
     DIAGNOSTIC_FLAG_BUT_LOCKED,
     EXPERIMENTAL_FLAG_BUT_LOCKED,
     DEVELOPER_FLAG_BUT_PRODUCT_BUILD,
-    NOTPRODUCT_FLAG_BUT_PRODUCT_BUILD
```
Comment (stefank) [inline src/hotspot/share/runtime/flags/jvmFlag.hpp:118]: Should the ',' on the previous line be removed?

File: src/hotspot/share/runtime/arguments.cpp (hunk 1 above)
Comment (coleenp) [reply src/hotspot/share/runtime/arguments.cpp:3420]: Yes, ifdef ASSERT is more appropriate.

File: src/hotspot/share/runtime/flags/jvmFlag.hpp (hunk 2 above)
Comment (coleenp) [reply src/hotspot/share/runtime/flags/jvmFlag.hpp:118]: Yes, I guess our compilers don't complain about that anymore.

**Review by coleenp — COMMENTED** [review]
Thanks for looking through the changes, Stefan.

**Review by vnkozlov — CHANGES_REQUESTED** [review]
So you finally decided to take on [JDK-8183288](https://bugs.openjdk.org/browse/JDK-8183288)

Essentially you are removing "optimized" VM build with these changes. In this case you need to change make files.

All Statistics flags should be product (which will increase product VM size) - it is important. May be need build's variable `--enable-jvm-feature-statistcs` to include statistics code on demand.

**Comment by coleenp** [comment]
The optimized build still works as before (actually surprised it still builds). Since for a long time the notproduct options acted like develop options, they still do just the same for the optimized build.

For optimized, all the develop and notproduct options are materialized.   Now just develop, not distinguishing notproduct from that.  The same code enabled in PRODUCT is still enabled.  I haven't looked at that removing optimized bug in a while.

**Comment by vnkozlov** [comment]
> For optimized, all the develop and notproduct options are materialized.

Okay, I see what you did here. You touched only flags declaration and did not `#ifndef PRODUCT` which guards statistics code, for example. Optimized VM build will get that code but will not include DEBUG_ONLY and `#ifdef ASSERT` guarded code.

So we still need to be careful when we use `#ifndef PRODUCT` and `#ifdef ASSERT`.

**Review by vnkozlov — APPROVED** [review]
Good.

**Comment by coleenp** [comment]
Yes, I had to remember what optimized did.  It gets all the options, but builds with optimization and doesn't turn on asserts.  I only removed the notproduct flag distinction since it hasn't been distinct for years and it's confusing what we actually wanted it to do.

**Comment by vnkozlov** [comment]
Long ago, before [JDK-8024545](https://bugs.openjdk.org/browse/JDK-8024545) we did not have not_product flags declared in product build. Only debug flags were declared as constant. We relied on that change since then.
That is why you may see the issue with not materialized flags in product build.

**Review by kimbarrett — APPROVED** [review]
Looks good.  One minor nit.

Also, it seems that develop and nonproduct (before this change) flags have
associated JVMFlag objects even in product builds. (The function
JVMFlag::is_constant_in_binary is evidence for this. I didn't dig through all
the code to verify it.) Probably if one were going to retain nonproduct
options and make them work "properly", they would only have JVMFlag objects in
non-product builds. But it's not obvious to me why we would want such objects
for either category in product builds. I think any change along that line
should be a separate followup.

**Comment by coleenp** [comment]
Thanks for reviewing, Kim.  Is your suggestion to not have a JVMFlag object for develop flags in PRODUCT builds? Presumably to save some footprint?  I'm not sure we would win fighting the macros to accomplish this.

**Comment by kimbarrett** [comment]
> Thanks for reviewing, Kim. Is your suggestion to not have a JVMFlag object for develop flags in PRODUCT builds? Presumably to save some footprint? I'm not sure we would win fighting the macros to accomplish this.

Yes, that's the suggestion and the rationale for it.  It should also remove the need for is_constant_in_binary.  I don't
know how hard it would actually be to accomplish this.  I agree it might not be worth the effort, but we won't know
until someone looks, which I haven't done.  It might even be easy.

**Comment by iklam** [comment]
> > Thanks for reviewing, Kim. Is your suggestion to not have a JVMFlag object for develop flags in PRODUCT builds? Presumably to save some footprint? I'm not sure we would win fighting the macros to accomplish this.
> 
> Yes, that's the suggestion and the rationale for it. It should also remove the need for is_constant_in_binary. I don't know how hard it would actually be to accomplish this. I agree it might not be worth the effort, but we won't know until someone looks, which I haven't done. It might even be easy.

Currently the VM prints an error message for non-product flags, so we need to keep some information about them. We can probably skip the type information, etc, to save a little space, but the space saving would be minimal.

```
$ java -XX:+LoomDeoptAfterThaw --version
Error: VM option 'LoomDeoptAfterThaw' is develop and is available only in debug version of VM.
Improperly specified VM option 'LoomDeoptAfterThaw'
Error: Could not create the Java Virtual Machine.
Error: A fatal exception has occurred. Program will exit.
```

**Pull request closed without merging**